src/cargo/ops/cargo_rustc/fingerprint.rs

```diff
@@ -474,6 +501,63 @@ pub fn prepare_build_cmd<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>)
     Ok((if compare.is_ok() {Fresh} else {Dirty}, write_fingerprint, Work::noop()))
 }
 
+fn build_script_local_fingerprints<'a, 'cfg>(cx: &mut Context<'a, 'cfg>,
+                                             unit: &Unit<'a>)
+    -> CargoResult<(Vec<LocalFingerprint>, Option<PathBuf>)>
+{
+    let state = cx.build_state.outputs.lock().unwrap();
+    // First up, if this build script is entirely overridden, then we just
+    // return the hash of what we overrode it with.
+    //
+    // Note that the `None` here means tha twe don't want to update the local
+    // fingerprint afterwards because this is all just overridden.
+    if let Some(output) = state.get(&(unit.pkg.package_id().clone(), unit.kind)) {
+        debug!("override local fingerprints deps");
+        let s = format!("overridden build state with hash: {}",
+                        util::hash_u64(output));
+        return Ok((vec![LocalFingerprint::Precalculated(s)], None))
+    }
+
+    // Next up we look at the previously listed dependencies for the build
+    // script. If there are none then we're in the "old mode" where we just
+    // assume that we're changed if anything in the packaged changed. The
+    // `Some` here though means that we want to update our local fingerprints
+    // after we're done as running this build script may have created more
+    // dependencies.
+    let deps = &cx.build_explicit_deps[unit];
+    let output = deps.build_script_output.clone();
+    if deps.rerun_if_changed.is_empty() && deps.rerun_if_env_changed.is_empty() {
+        debug!("old local fingerprints deps");
+        let s = pkg_fingerprint(cx, unit.pkg)?;
+        return Ok((vec![LocalFingerprint::Precalculated(s)], Some(output)))
+    }
+
+    // Ok so now we're in "new mode" where we can have files listed as
+    // dependencies as well as env vars listed as dependencies. Process them all
+    // here.
+    Ok((local_fingerprints_deps(deps, unit.pkg.root()), Some(output)))
+}
+
+fn local_fingerprints_deps(deps: &BuildDeps, root: &Path) -> Vec<LocalFingerprint> {
+    debug!("new local fingerprints deps");
+    let mut local = Vec::new();
+    if !deps.rerun_if_changed.is_empty() {
+        let output = &deps.build_script_output;
+        let deps = deps.rerun_if_changed.iter().map(|p| root.join(p));
+        let mtime = mtime_if_fresh(output, deps);
+        let mtime = MtimeSlot(Mutex::new(mtime));
+        local.push(LocalFingerprint::MtimeBased(mtime, output.clone()));
+    }
+
+    for var in deps.rerun_if_env_changed.iter() {
```

Comment:
But there's a slim chance that buildscript itself prints them non-deterministically (like iterating via HashMap). But yeah, that probably doesn't matter much.